### PR TITLE
Fix optparse fallback logic in Django 1.8+

### PR DESCRIPTION
**Pull request by jrdietrick**
**Disclaimer**: I have read the discussions in https://github.com/jezdez/django-configurations/issues/105#issuecomment-88895512 and https://github.com/jezdez/django-configurations/commit/0ff4c7612d28036449edd4b2cf3b58ca1eded0f2, and I understand the situation this package is in. I pondered long and hard as to whether I should even open this pr.

I completely get it; by the time I am done: root-causing this issue, learning how configurations works, learning what the Django guys were trying to do in 1.8, learning how to run tox, writing tests, and finally authoring this pr, I will have spent _many hours_ on these three lines of diff. Hours I could have spent on "real work". I'm just fortunate -- and love (and need) this package enough! -- that my current workload happens to be light enough, and I happen to have enough caffeine, to make it happen. I fully understand that it could take hours more of time that someone doesn't have to confidently sign off on it.

**tl;dr:** If there's no time to even read this pull req, let alone merge it, no hard feelings!

----

OK, so here's the deal:

Django switched from optparse to argparse in Django 1.8 for extending management commands. To ease deprecation, until Django 1.10, there's some fun shim logic in Django. It says, "If the command I'm running appends anything to its `option_list` (the old way), we'll still use optparse. Otherwise, we'll use argparse."

Note that, until Django 1.10, the old optparse behavior gets checked first. In other words, if `option_list` is manipulated _and_ `add_arguments` (the new way) is implemented, we still use optparse, and anything in `add_arguments` is ignored.

That's important for configurations, because it means that we can't just, for instance, add our `--configuration` option to `option_list` up front and deal with the problem when it really gets deprecated in Django 1.10. In order to correctly wrap management commands from arbitrary packages, which may or may not have shifted over to `add_arguments` yet, we have to _wait and see what the command does_ (and thus what parser we get back), and then play our hand.

https://github.com/jezdez/django-configurations/commit/6ce3740365d92c329c1e20036266cd090d57b7c4 made an attempt at doing this, but the `base.BaseCommand.option_list += configuration_options` causes two problems, actually:

1. It doesn't actually add the configuration option to the command we are running right now. We just got handed the parser, which already looked at `BaseCommand.option_list`. So this line only adds the option for any commands run _after_ this. This is the subject of the test in the second commit of this pr, which fails on Django 1.8 if run without the patch in the other commit.

2. We're touching `base.BaseCommand.option_list`, which is a class variable. If we now run another management command in the same Django instance (say, via `call_command`), we will add the `--configuration` option a second time. Do this enough times (three, exactly), and you'll get an `OptionConflictError`. I encountered this while using django-nose, which evidently runs some subcommands under its `test` command. This is the cause of #109, and pr #110 implements a slightly hacky but working fix, but doesn't fix problem 1 above.

The proposed patch fixes both issues by:
* maintaining the existing "wait and see" behavior that accommodates both mgmt commands which have and have not upgraded their argument extension
* handling the optparse fallback case by one-off adding our argument(s) to the parser in our wrapped routine, leaving the global `BaseCommand.option_list` untouched

P.S. - Using the three management commands below and then running `manage.py command1` on Django 1.8 is enough to minimally repro the `OptionConflictError` problem. It would not be difficult to write a test to validate this doesn't happen any longer, but seemed like overkill.

```python
# command1.py
class Command(BaseCommand):

    option_list = BaseCommand.option_list + (
        make_option('--arg1', action='store_true'),
    )

    def handle(self, *args, **options):
        call_command('command2')

# command2.py
class Command(BaseCommand):

    option_list = BaseCommand.option_list + (
        make_option('--arg2', action='store_true'),
    )

    def handle(self, *args, **options):
        call_command('command3')

# command3.py
class Command(BaseCommand):

    option_list = BaseCommand.option_list + (
        make_option('--arg3', action='store_true'),
    )

    def handle(self, *args, **options):
        print 'hi'
```

### CODE REVIEW

**Comment by blueyed** [comment]
Thanks a lot for this!

**Comment by skolsuper** [comment]
@blueyed Do you think this branch is the best choice for a >0.8 release going forward?

@everyone: Since some people (myself included) appear to have been using this package in production already, shall we think about making a 1.0 release and what needs to be done for that to happen?

**Comment by blueyed** [comment]
@skolsuper 
Not sure.  At least it should be updated to include py35, and then I think e.g. https://github.com/jazzband/django-configurations/pull/128 is better (has more tests), but misses the bumping of the requirement.

**Comment by blueyed** [comment]
@skolsuper 
I think we should get the tests fixed / updated first.
I've updated/rebased https://github.com/jazzband/django-configurations/pull/128.

**Comment by blueyed** [comment]
#131 is now about fixing tests and the Travis setup, and is ready for review.

**Comment by jrdietrick** [comment]
Just rebased this off of the latest jazzband/master (which includes #131 to fix tests), so this PR is now solely about the functional issue in importer.py again (addresses #109, #120).

**Comment by blueyed** [comment]
:+1:

**Comment by mauricioabreu** [comment]
Is unittest2 only necessary for Python 2.6 right?

**Comment by skolsuper** [comment]
Yes @jrdietrick you can get the `@skipIf` decorator from `django.utils.unittest` I believe.

**Comment by jrdietrick** [comment]
@skolsuper yes, but it seems in Python 2.7+ only, as @mauricioabreu mentions ([docs](https://docs.python.org/2/library/unittest.html#skipping-tests-and-expected-failures)). So, unless I'm mistaken, our Travis builds on py26 will fail. What do you recommend?

**Comment by skolsuper** [comment]
`django.utils.unittest` backports unittest2 for python 2.5+. [Link](https://docs.djangoproject.com/en/1.5/topics/testing/overview/#writing-tests)

**Comment by jrdietrick** [comment]
Oh, right you are. Sorry, I should have read your original comment more carefully.

Just to make it a little more complicated, `django.utils.unittest` is removed in Django 1.9, so our import is wrapped in an `if` now to get the right one.

I've updated the PR with one more commit.

**Comment by mauricioabreu** [comment]
LGTM :+1: 
When importing packages that may not be installed you can take advantage of `try...except` feature.

**Comment by ticosax** [comment]
:+1:

**Comment by blueyed** [comment]
:+1:
